src/assignment-2/BasicVault.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "yield-utils-v2/token/IERC20.sol";
+
+/// @title Vault holding ERC20 tokens for users
+/// @author davidbrai
+/// @notice The vault allows users to deposit and withdraw token of a specific ERC20 contract. The contracts keeps track of balances.
+contract BasicVault {
+
+    /// @notice ERC20 token which this vault holds tokens for
+    IERC20 immutable public token;
+
+    /// @notice Mapping from address to balance representing the balance of each user
+    mapping(address => uint) public balances;
+
+    error BalanceTooLow();
+    error TransferFailed();
+
+    /// @notice This event is emitted when a user deposits tokens
+    event Deposit(address from, uint amount);
+
+    /// @notice This event is emitted when a user withdraws tokens
+    event Withdraw(address to, uint amount);
+
+    /// @notice Initalizes a new vault
+    /// @param _token The address of an ERC20 token contract
+    constructor(IERC20 _token) {
+        token = _token;
+    }
+
+    /// @notice Deposits tokens from the user into the vault
+    /// @dev Updates the balances mapping with the amount of tokens
+    /// @param amount The amount of tokens to deposit
+    function deposit(uint amount) public {
+        bool success = token.transferFrom(msg.sender, address(this), amount);
```

Comment:
When coding Yield v1 I did it this way, instead of using checks-effects-interactions. Basically you put the lines that are beneficial to the user last.

It is easier to understand, and that's why I did it. However, as complexity grows it becomes harder and harder to see what is safe and what is not. Checks-effects-interactions saves you from having to think about what actions are user-positive and in which order, and is safest in the long term.

Also, you will throw auditors off the scent if you don't use checks-effects-interactions, and even if you did it safely, they will spend extra time looking for issues there, instead of where you want them to be.

Reply:
makes sense! thanks for the thoughtful feedback. I will change it to use the checks-effects-interactions (CEI ?) pattern.

Reply:
fixed here: https://github.com/davidbrai/yield-mentorship/pull/2/commits/fdc57876424aa9961a661227d7d231b4d9593f19